crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,642 @@
+                let mut packed = self.pack(py, value)?;
+                std::mem::swap(&mut packed, &mut self.data[index]);
+                Ok(())
+            }
+        }
+    }
+
+    pub fn insert(
+        &mut self,
+        py: Python<'_>,
+        index: isize,
+        value: PyRef<CircuitInstruction>,
+    ) -> PyResult<()> {
+        let index = self.convert_py_index_clamped(index);
+        let packed = self.pack(py, value)?;
+        self.data.insert(index, packed);
+        Ok(())
+    }
+
+    pub fn pop(&mut self, py: Python<'_>, index: Option<PyObject>) -> PyResult<PyObject> {
+        let index =
+            index.unwrap_or_else(|| std::cmp::max(0, self.data.len() as isize - 1).into_py(py));
+        let item = self.__getitem__(py, index.as_ref(py))?;
+        self.__delitem__(py, index.as_ref(py).extract()?)?;
+        Ok(item)
+    }
+
+    pub fn append(&mut self, py: Python<'_>, value: PyRef<CircuitInstruction>) -> PyResult<()> {
+        let packed = self.pack(py, value)?;
+        self.data.push(packed);
+        Ok(())
+    }
+
+    // To prevent the entire iterator from being loaded into memory,
+    // we create a `GILPool` for each iteration of the loop, which
+    // ensures that the `CircuitInstruction` returned by the call
+    // to `next` is dropped before the next iteration.
+    pub fn extend(&mut self, py: Python<'_>, itr: &PyAny) -> PyResult<()> {
+        // To ensure proper lifetime management, we explicitly store
+        // the result of calling `iter(itr)` as a GIL-independent
+        // reference that we access only with the most recent GILPool.
+        // It would be dangerous to access the original `itr` or any
+        // GIL-dependent derivatives of it after creating the new pool.
+        let itr: Py<PyIterator> = itr.iter()?.into_py(py);
+        loop {
+            // Create a new pool, so that PyO3 can clear memory at
+            // the end of the loop.
+            let pool = unsafe { py.new_pool() };
+
+            // It is recommended to *always* immediately set py to the pool's
+            // Python, to help avoid creating references with invalid lifetimes.
+            let py = pool.python();
+
+            // Access the iterator using the new pool.
+            match itr.as_ref(py).next() {
+                None => {
+                    break;
+                }
+                Some(v) => {
+                    self.append(py, v?.extract()?)?;
+                }
+            }
+            // The GILPool is dropped here, which cleans up the ref
+            // returned from `next` as well as any resources used by
+            // `self.append`.
+        }
+        Ok(())
+    }
+
+    pub fn clear(&mut self, _py: Python<'_>) -> PyResult<()> {
+        std::mem::take(&mut self.data);
+        Ok(())
+    }
+
+    // Marks this pyclass as NOT hashable.
+    #[classattr]
+    const __hash__: Option<Py<PyAny>> = None;
```

Comment:
This feels like a slightly weird setup that it's an `Option<Py<PyAny>>` and not directly a `Python::None()`.  Or can we not get one of those, because we're not in an execution context during module initialisation?

Reply:
^that, I believe. I believe a newer version of PyO3 adds `Python::None` as a proper type, but I'm not sure this is worth messing with since it's taken verbatim from the PyO3 docs.